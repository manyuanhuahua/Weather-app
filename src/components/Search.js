import React, { useState, useEffect } from "react";

import { useHistory } from "react-router-dom";
import "./style/search.css"

import { KEY, URL } from "./Api.js";

const Search = () => {
  // const [select, setSelect] = useState(0);
  // const [selectValue, setSelectValue] = useState("");
  const history = useHistory();
  const [city, setCity] = useState(null)
  const [state, setState] = useState(null)
  const [searchRes, setSearchRes] = useState({})
  const [searchLoad, setSearchLoad] = useState(false);
  const [noData,setNoData] = useState('')


// convert user's input of city and state into geo coordinates and fetch weather info based on coordinates
  const converGeo = (city, state) => {
    fetch(
      `http://api.openweathermap.org/geo/1.0/direct?q=${city},${state},US&limit=1&appid=${KEY}`
    )
      .then((res) => res.json())
      .then((data) => {
        if(!data.length){
          setNoData('Location not found!Please try another one!')
          return
        }else{
          const lat = data[0].lat
          const lng = data[0].lon
          setSearchRes({});
          fetch(
            `${URL}&lat=${lat}&lon=${lng}&units=metric&cnt=40&exclude=hourly,minutely&appid=${KEY}`
          ).then((res) => res.json())
            .then((data) => {

              setSearchRes(data);
              setSearchLoad(true)
          })

  }
});

  }


  const handlesubmit = (e) => {
    e.preventDefault()
    converGeo(city, state)
  }


// update history state based on input and redirect user to weatherlist page
  useEffect(() => {
    if (searchLoad && Object.keys(searchRes).length) {
      history.push({
        pathname: `/weatherlist`,
        state: { searchRes }
      });
      window.location.reload();
    }
  }, [searchLoad, history, searchRes]);
  
  return (
    <div className="search-container">
      <form onSubmit={handlesubmit}>
        <div className="row">
          <div className="col-md-3">
            <input
              type="text"
              className="form-control"
              id="city"
              name="city"
              placeholder="City"
              onChange={(e) =>{
                setNoData('')
                setCity(e.target.value.toLowerCase())}}
              required
              autoComplete='false'
            ></input>
          </div>
          <div className="col-md-3">
            <input
              type="text"
              className="form-control"
              name="state"
              id="state"
              placeholder="State"
              onChange={(e) =>{
                setState(e.target.value.toLocaleLowerCase())
                setNoData('')
              }}
              required={true}

            ></input>
          </div>
          <div className="col-md-3 mt-md-0 text-md-left" style={{fontWeight:'bold'}}>
            <button className="btn btn-warning" style={{fontWeight:'600'}}>Get Weather</button>
          </div>
        </div>
      {noData !== '' &&
        <p style={{textAlign:'center',color:'#ac3f21',marginBottom:'0'}}>{noData}</p>
      }
      </form>
    </div>
  );
};

export default Search;
