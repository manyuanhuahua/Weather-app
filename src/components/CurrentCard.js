import moment from "moment";
import React, { useEffect, useState } from "react";
import "./style/summarycard.css";
import { KEY, ICON } from "./Api.js";
import "./style/current.css";
import Search from "./Search";



const CurrentCard = () => {
  const defaultIcon = `http://openweathermap.org/img/wn/02d@2x.png`;
  const [curWeatherIcon, setCurWeatherIcon] = useState(defaultIcon);
  const [time, setTime] = useState(moment().format("LT"));
  const [humidity, setHumidity] = useState(null);
  const [pressure, setPressure] = useState(null);
  const [feelLike, setFeelLike] = useState(null);
  const [description, setDescription] = useState(null);
  const [sunrise, setSunsire] = useState(null);
  const [sunset, setSunset] = useState(null);
  const [load, setLoad] = useState(false);
  const [img, setImg] = useState(null);

  const [curLoc, setCurLoc] = useState(null);
  const [timeZone, setTimeZone] = useState(null);
  const [country, setCountry] = useState(null);
  const [title, setTitle] = useState();



// update clock time every minute
  useEffect(() => {
    const interval = setInterval(() => {
      const cur = new Date();
      moment.locale();
      setTime(moment(cur).format("LT"));
    }, 60000);
    return () => clearInterval(interval);
  }, []);

// get current location and fetch local weather info
  const getCurLocation = () => {
    navigator.geolocation.getCurrentPosition((success) => {
      let { latitude, longitude } = success.coords;

      fetch(
        `https://api.openweathermap.org/data/2.5/weather?lat=${latitude}&lon=${longitude}&units=metric&appid=${KEY}`
      )
        .then((res) => res.json())
        .then((data) => {
          showWeather(data);
          setLoad(true);
        })

    });
  };

  // get daily astronomy image from NASA
  const image = () => {
    fetch("https://go-apod.herokuapp.com/apod")
      .then((res) => res.json())
      .then((data) => {
        setImg(data.url)
        setTitle(data.title)

      })
  }

  image()


// display local weather info
  const showWeather = (data) => {
    let { humidity, pressure, feels_like } = data.main;
    let { description, icon } = data.weather[0];
    let { sunrise, sunset, country } = data.sys;

    setTimeZone(country)
    setHumidity(humidity);
    setPressure(pressure);
    setFeelLike(parseInt(feels_like));
    setDescription(description);
    setSunsire(moment(sunrise * 1000).format("HH:mm a"));
    setSunset(moment(sunset * 1000).format("HH:mm a"));
    setCurWeatherIcon(`${ICON + icon}@2x.png`);
    setCurLoc(data.name);
    setCountry(data.coord.lat + 'N' + data.coord.lon + 'E');

  };


  getCurLocation();


  return (
    <div className="cur-container">
      <div className="current-info">
        <div className="current-date-contianer">
          <div className="time" id="time">
            {time}
          </div>
          <div className="date" id="date">
            {moment(new Date()).format("dddd L")}
          </div>
          <div className="location">{curLoc}</div>
          <div className="others" id="current-weather-item">
            {load ? (
              <>
                <div className="icon">
                  <div className="weather-item">
                    <img src={curWeatherIcon} alt="" />
                    <div style={{ margin: "auto", fontSize: "30px", color: '#e7b3b3' }}>
                      {description}
                    </div>
                  </div>
                </div>
                <div className="weather-item">
                  <div>Humidity</div>
                  <div>{humidity}%</div>
                </div>
                <div className="weather-item">
                  <div>Pressure</div>
                  <div>{pressure} hPa</div>
                </div>
                <div className="weather-item">
                  <div>FeelLike</div>
                  <div>{feelLike}&deg;C</div>
                </div>

                <div className="weather-item">
                  <div>Sunrise</div>
                  <div>{sunrise}</div>
                </div>
                <div className="weather-item">
                  <div>Sunset</div>
                  <div>{sunset}</div>
                </div>
              </>
            ) : (
              <>
                <h1>Loading Data</h1>
              </>
            )}
          </div>
        </div>
        <div className="place-container">
          <div className="place-top">
            <div className="time-zone">{timeZone}</div>
            <div id="country" className="country">
              {country}
            </div>

          </div>
          <div className="search-bar" style={{ marginTop: '1rem' }}>
            <Search />
          </div>
          <div className="img-container">
            <h2>Astronomy For Today</h2>
            {img ? <img src={img} alt='' style={{ width: '400px', height: '300px', objectFit: 'cover' }} /> : <p>Loading</p>}
            <h4 style={{ width: '400px', overflowWrap: 'break-word', color: '#6e3b3b' }}>{title}</h4>
          </div>
        </div>

      </div>
    </div>
  );
};

export default CurrentCard;
