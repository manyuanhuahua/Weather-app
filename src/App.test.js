import React from "react";
import moment from "moment";

import { shallow } from "enzyme";
import { render, fireEvent } from '@testing-library/react';
import CurrentCard from "./components/CurrentCard";
import Enzyme from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';

import Search from './components/Search';





Enzyme.configure({ adapter: new Adapter() });

// navigator.geolocation.getCurrentPosition is not working in the test environment,
// so mock the geolocation function and provide a fake implementation

global.navigator.geolocation = {
  getCurrentPosition: jest.fn().mockImplementation((success) => {
    let fakeData = {
      coords: {
        latitude: 37.7749,
        longitude: -122.4194
      }
    };
    success(fakeData);
  })
};


// homepage testing
describe("CurrentCard component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<CurrentCard />);
  });

  it('renders the component', () => {
    expect(wrapper).toHaveLength(1);
  });

  it('renders the loading message', () => {
    expect(wrapper.find('h1').text()).toEqual('Loading Data');
  });



},)


// search component

describe('Search component', () => {
  let wrapper;

  let searchLoad, searchRes;

  beforeEach(() => {


    wrapper = shallow(<Search />);
    searchLoad = true;
    searchRes = { city: 'New York' };

  });

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update the state on input change', () => {
    const { getByPlaceholderText } = render(<Search />);
    const cityInput = getByPlaceholderText('City');
    fireEvent.change(cityInput, { target: { value: 'San Francisco' } });
    expect(cityInput.value).toBe('San Francisco');

    const stateInput = getByPlaceholderText('State');
    fireEvent.change(stateInput, { target: { value: 'California' } });
    expect(stateInput.value).toBe('California');
  });

  it('calls the API on form submit', async () => {
    const mockJsonPromise = Promise.resolve({});
    const mockFetchPromise = Promise.resolve({ json: () => mockJsonPromise });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

    wrapper.find('form').simulate('submit', { preventDefault: jest.fn() });

    expect(global.fetch).toHaveBeenCalled();
    await expect(mockFetchPromise).resolves.toBeDefined();
  });

})
