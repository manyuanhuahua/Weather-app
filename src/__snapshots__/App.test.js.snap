// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders without crashing 1`] = `ShallowWrapper {}`;
